libraries/ui/src/main/java/androidx/media3/ui/SubtitlePainter.java

```diff
@@ -470,4 +472,67 @@ private static boolean areCharSequencesEqual(
     // equals methods, so we perform one explicitly here.
     return first == second || (first != null && first.equals(second));
   }
+
+  /**
+   * Checks whether the given {@link CharSequence} contains any characters
+   * with right-to-left (RTL) directionality.
+   * <p>
+   * This method inspects each character's Unicode directionality and returns {@code true}
+   * if at least one character is classified as RTL. This includes characters with the following
+   * directionalities:
+   * <ul>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT}</li>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC}</li>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT_EMBEDDING}</li>
+   *     <li>{@link Character#DIRECTIONALITY_RIGHT_TO_LEFT_OVERRIDE}</li>
+   * </ul>
+   *
+   * @param input the input {@link CharSequence} to analyze
+   * @return {@code true} if the input contains at least one RTL character; {@code false} otherwise
+   */
+  private static boolean containsRTL(@Nullable CharSequence input) {
+    if (input == null) {
+      return false;
+    }
+    for (int i = 0; i < input.length(); i++) {
+      char ch = input.charAt(i);
+      byte dir = Character.getDirectionality(ch);
```

Comment:
I think it would be better/more robust to deal in code points here, so it all works with supplementary plane characters too.

That probably means either:
* Gate this logic on API 24 and use `String.codePoints()`
* Or to support API 21+ use `String.codePointCount()` as your loop control length, with careful incrementing of `i` by 1 or 2 depending on whether you read a surrogate pair with `String.codePointAt` or not.
     * Beware that we are likely moving the whole library to min API 23 relatively soon, so then it just becomes "to support API 23"